src/Highway.Data/Repositories/Repository.cs

```diff
@@ -9,26 +8,18 @@ namespace Highway.Data
     ///     A Entity Framework Specific repository implementation that uses Specification pattern to execute Queries in a
     ///     controlled fashion.
     /// </summary>
-    public class Repository : IRepository
+    public class Repository : ReadonlyRepository, IRepository
```

Comment:
This should probably have a common base type with ReadonlyRepository, rather than inherit directly from it.  The reason is that through these constructors, it is possible for a ReadonlyRepository to wind up with a writable data context.  Thoughts?

Reply:
This one is rough for me, part of me wants to split the interface into read and write operations. Another portion likes the construction safety of specific context variables. If the context configuration is used in DataContext, could we surface that to adjust the interactions of the repository?

Reply:
@DevlinLiles I think the IReadonlyRepository and associated ReadonlyRepository are better choices.  At his level, registration could get quite complex, and I might have difficulty making a sufficient change without changing the public surface of the involved types.

I really like the declarative intent of a dependency asking for the IReadonlyRepository service.  That seems extremely cut and try.  How does a service indicate that it needs an IRepository backed by an IReadonlyDataContext or an IDataContext configured for readonly?

I think the configurable DataContext makes much more sense than a configurable Repository.

Any more thoughts?

Reply:
I agree I like the declarative intent at the repository, and the configuration option at the context level. Adding a constructor that takes readonly, and a gated check for write operations would do this nicely. We have to be careful to add a constructor rather than a defaulted additional parameter for any that have inherited. The behavior should stay the same for them.

Reply:
> We have to be careful to add a constructor rather than a defaulted additional parameter for any that have inherited. The behavior should stay the same for them.

Is this true?  Does the compiler differentiate between a default additional parameter and a new constructor?  I don't think using a defaulted parameter will break the behavior, so long as the new parameter's default matches the original behavior.  I don't think old calls to the constructor will have behavioral changes.

Reply:
You may be right, I would have to double check. Just want to make sure. Potentially test this before change / after from an inheriting class?

Reply:
@DevlinLiles I looked it up.  The compiler creates something different, exposing the new parameter AND its default value.  The caller (or developer in an IDE) will see that new parameter and its default value.  The additional overloads are just that, and do NOT declaratively expose the default behavioural intent.

I think I prefer the new parameter, because of that additional bit of discoverability.  However, we should consider whether this might be being called from languages or interop runtimes that do not support optional parameters.  Some research is necessary around that, as I don't know if it is a real thing these days, but John Skeet said it was 10 years ago (or whatever that was), so I'll take a look.